pallets/vrf/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+	pub trait GetMostRecentVrfInputs<RelayHash, SlotNumber> {
+		fn get_most_recent_relay_block_hash() -> RelayHash;
+		fn get_most_recent_relay_slot_number() -> SlotNumber;
+	}
+
+	/// This trait tells us if the round changed in the current block
+	/// => new set of authorities to be put in storage
+	pub trait RoundChangedThisBlock {
+		fn round_changed_this_block() -> bool;
+	}
+
+	/// Exposes randomness in this pallet to the runtime
+	pub trait GetRandomness {
+		type Randomness;
+		fn get_last_randomness() -> Self::Randomness;
+		fn get_current_randomness() -> Self::Randomness;
+	}
+
+	/// Make VRF transcript
+	pub fn make_transcript<Hash: AsRef<[u8]>>(
+		relay_slot_number: Slot,
+		relay_block_hash: Hash,
+	) -> Transcript {
+		let mut transcript = Transcript::new(&BABE_ENGINE_ID);
+		transcript.append_u64(b"relay slot number", *relay_slot_number);
+		transcript.append_message(b"relay block hash", relay_block_hash.as_ref());
+		transcript
+	}
+
+	/// Pallet for storing and exposing VRF outputs by block authors
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The relay block hash type (probably H256)
+		type RelayBlockHash: Parameter
+			+ Member
+			+ MaybeSerializeDeserialize
+			+ Default
+			+ Copy
+			+ AsRef<[u8]>;
+		/// Gets the most recent relay block hash and relay slot number in `on_initialize`
+		type MostRecentVrfInputGetter: GetMostRecentVrfInputs<Self::RelayBlockHash, Slot>;
+		/// Convert account to VRF key, presumably via AuthorMapping instance
+		/// TODO: maybe TryConvert and map to error because fallible conversion
+		type AccountToVrfId: Convert<Self::AccountId, AuthorityId>;
+		/// Whether or not the round changed this block
+		type RoundChanged: RoundChangedThisBlock;
+		/// Get the selected candidate accounts from staking
+		type SelectedCandidates: Get<Vec<Self::AccountId>>;
+	}
+
+	/// Current block randomness
+	/// Set in `on_initialize`, before it will contain the randomness from the last block
+	#[pallet::storage]
+	#[pallet::getter(fn current_randomness)]
+	pub type CurrentRandomness<T> = StorageValue<_, MaybeRandomness, ValueQuery>;
+
+	/// Last block randomness
+	/// Set in `on_initialize`, before it will contain the randomness from the last last block
+	#[pallet::storage]
+	#[pallet::getter(fn last_randomness)]
+	pub type LastRandomness<T> = StorageValue<_, MaybeRandomness, ValueQuery>;
+
+	/// Current set of authorities by AuthorityId
+	/// Set in `on_initialize` upon round changes
+	#[pallet::storage]
+	#[pallet::getter(fn authorities)]
+	pub type Authorities<T> = StorageValue<_, Vec<AuthorityId>, ValueQuery>;
+
+	/// Most recent relay chain block hash
+	/// Set in `on_initialize` before setting randomness
+	#[pallet::storage]
+	#[pallet::getter(fn most_recent_relay_block_hash)]
+	pub(crate) type MostRecentRelayBlockHash<T: Config> =
```

Comment:
Are MostRecentSlotNumber and MostRecentRelayBlockHash mostly read together? if So we can consider making this a signle storage Item, called RelayBeacon or something similar

Reply:
They are mostly read/written together so I will follow this suggestion

Reply:
https://github.com/PureStake/moonbeam/pull/1376/commits/b1635bbe2085ffa8a4b4f405b2bfcc3184ca2e95 resolves